clients/src/main/java/org/apache/kafka/common/record/FileRecords.java

```diff
@@ -0,0 +1,123 @@
+    private final File file;
+    private final FileChannel channel;
+    private final long start;
+    private final long end;
+    private final long size;
+
+    public FileRecords(File file,
+                       FileChannel channel,
+                       int start,
+                       int end,
+                       boolean isSlice) throws IOException {
+        this.file = file;
+        this.channel = channel;
+        this.start = start;
+        this.end = end;
+
+        if (isSlice)
+            this.size = end - start;
+        else
+            this.size = Math.min(channel.size(), end) - start;
+    }
+
+    @Override
+    public int sizeInBytes() {
+        return Long.valueOf(size).intValue();
+    }
+
+    @Override
+    public long writeTo(GatheringByteChannel destChannel, long offset, int length) throws IOException {
+        long newSize = Math.min(channel.size(), end) - start;
+        if (newSize < size)
+            throw new KafkaException(String.format("Size of FileRecords %s has been truncated during write: old size %d, new size %d", file.getAbsolutePath(), size, newSize));
+
+        if (offset > size)
+            throw new KafkaException(String.format("The requested offset %d is out of range. The size of this FileRecords is %d.", offset, size));
+
+        long position = start + offset;
+        long count = Math.min(length, this.size - offset);
+        if (destChannel instanceof TransportLayer) {
+            TransportLayer tl = (TransportLayer) destChannel;
+            return tl.transferFrom(this.channel, position, count);
+        } else {
+            return this.channel.transferTo(position, count, destChannel);
+        }
+    }
+
+    @Override
+    public RecordsIterator iterator() {
+        return new RecordsIterator(new FileLogInputStream(channel, start, end), false);
+    }
+
+    private static class FileLogInputStream implements LogInputStream {
+        private long position;
+        protected final long end;
+        protected final FileChannel channel;
+        private final ByteBuffer logHeaderBuffer = ByteBuffer.allocate(Records.LOG_OVERHEAD);
+
+        public FileLogInputStream(FileChannel channel, long start, long end) {
+            this.channel = channel;
+            this.position = start;
+            this.end = end;
+        }
+
+        @Override
+        public LogEntry nextEntry() throws IOException {
+            if (position + Records.LOG_OVERHEAD >= end)
+                return null;
+
+            logHeaderBuffer.rewind();
+            channel.read(logHeaderBuffer, position);
+            if (logHeaderBuffer.hasRemaining())
+                return null;
+
+            logHeaderBuffer.rewind();
+            long offset = logHeaderBuffer.getLong();
+            int size = logHeaderBuffer.getInt();
+            if (size < 0)
+                throw new IllegalStateException("Record with size " + size);
+
+            ByteBuffer recordBuffer = ByteBuffer.allocate(size);
```

Comment:
This already exists in FileMessageSet, but perhaps we could avoid reading from the channel if end - position is less than size?